### PR TITLE
chore(architecture): refactor PathIter to avoid regex + allocations (plus a bonus change)

### PR DESCRIPTION
**Pull request by tobz**
We've reworked `PathIter` to avoid allocations where possible, as well as using the regular expression "fast path".

In profiling when investigating an issue reported by a user -- unexpectedly high memory usage (RSS specifically) -- it was determined that the bulk of allocations in their workload came from using the `remap` transform.  This was due to the nature of `PathIter`, allocating strings as it parsed paths, as well as running a regular expression over the input path to determine if a "fast path" could be taken.

We've reworked `PathIter` to remove this regular expression, as well as the fast path.  Instead, we parameterized `PathIter` over the lifetime of the input path, such that we can optimally return borrowed substrings when possible, only falling back to allocating when skipping characters is required, such as within escaped sequences.

A benchmark was written to demonstrate the performance gains, as evidenced below:

```
vector_core::event::util::log/PathIterOld (flat)                                                                            
                        time:   [98.540 ns 98.722 ns 98.930 ns]
vector_core::event::util::log/PathIterOld (nested)                                                                            
                        time:   [172.81 ns 173.00 ns 173.21 ns]
vector_core::event::util::log/PathIterOld (nested array)                                                                            
                        time:   [293.00 ns 293.58 ns 294.45 ns]
vector_core::event::util::log/PathIterOld (nested escaped)                                                                            
                        time:   [163.77 ns 164.12 ns 164.52 ns]

vector_core::event::util::log/PathIterNew (flat)                                                                            
                        time:   [57.294 ns 57.636 ns 57.993 ns]
vector_core::event::util::log/PathIterNew (nested)                                                                            
                        time:   [106.08 ns 106.26 ns 106.45 ns]
vector_core::event::util::log/PathIterNew (nested array)                                                                            
                        time:   [122.74 ns 123.29 ns 123.88 ns]
vector_core::event::util::log/PathIterNew (nested escaped)                                                                            
                        time:   [77.494 ns 77.729 ns 77.981 ns]
```

The results are fairly impressive: at the very low end, the new version is 60% faster, with some tests showing a staggering improvement of nearly 250%.  As well, we've significantly reduced allocations.  In the case of a path without escaped sequences, we perform no extra allocations in order to return the path components.  In the case of escaped sequences, we allocate a single `String` to serve as temporary storage, which likely will grow once or twice depending on the overall path length, but is reused for subsequent components.

As further evidence, this change was able to help reduce the RSS usage of aforementioned use case by around 15-20%.

## Bonus change: more efficient encoding for the `vector` v2 sink

We've also improved the efficiency of the `vector` v2 sink by configuring `prost` to use `Bytes` as the underlying storage for for the `raw_bytes` field when encoding `Value::Bytes`.  This lets us avoid an allocation to create a `Vec<u8>`, since we already use `Bytes` as the internal representation.

Signed-off-by: Toby Lawrence <toby@nuclearfurnace.com>

### CODE REVIEW

**Review by blt — APPROVED** [review]
Very nice work here.

File: src/transforms/remap.rs [hunk 1]
```diff
@@ -122,9 +136,10 @@ impl FunctionTransform for Remap {
 
         let mut target: VrlTarget = event.into();
 
-        let mut runtime = Runtime::default();
-
-        let result = runtime.resolve(&mut target, &self.program, &self.timezone);
+        self.runtime.clear();
+        let result = self
+            .runtime
+            .resolve(&mut target, &self.program, &self.timezone);
```
Comment (JeanMertz) [inline src/transforms/remap.rs:142]: It's probably a good idea to add a test to validate that this keeps working as expected (e.g. given two events, the runtime only keeps its state around per event)

**Review by JeanMertz — CHANGES_REQUESTED** [review]
Awesome change, thanks for tracking this down. I left one comment about adding a crucial test, but otherwise this LGTM.

File: src/transforms/remap.rs (hunk 1 above)
Comment (tobz) [reply src/transforms/remap.rs:142]: I added a simplistic test that I think at least confirms that variable assignments don't cross the invocation boundary?

**Pull request merged**